### ISSUE TITLE
how to access to collection path in `pep_api_gen_query_pre`?

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#7900]
### iRODS Version, OS and Version
4.3.2, almalinux8

### What did you try to do?
I would like to monitor queries executed for a given collection. Therefore I want to somehow fetch the collection path when the policy is triggered by a user's query action.

### Expected behavior
Expected to see a key like `objPath` or `collPath`.

### Observed behavior (including steps to reproduce, if applicable)
I am not able to access to relevant key to obtain the collection path.

I can see the following information see in the audit plugin output these attributes contain the collection path though.
`where_502":"='/u0137480/home/projectA'"`
or `where_501` but not sure in which circumstance that would contain a value.

**Comment by trel** [comment]
Queries don't have to be associated with a logical path.

It depends on the columns requested as part of the Query itself.

**Comment by mstfdkmn** [comment]
Okay got it. But I would like to find a way/hook to be invoked by any queries (from every client such as ils, walk() and the genquery utility) executed for a collection in order to implement a policy. As far as I tested/experienced I couldnt find one that is triggered by a query. That is why I tried pep_api_gen_query_pre. For example as far as names suggest, one of the following PEPs should be invoked when there is a query against a collection. 

```
pep_api_open_collection_pre
pep_api_query_spec_coll_pre
pep_api_read_collection_pre
```

Could you show me one that can be used in this context?

**Comment by korydraughn** [comment]
All GenQuery queries go to the database so you may want to try the GenQuery database PEPs - `pep_database_gen_query_*`.

**Comment by mstfdkmn** [comment]
I had tested this earlier but observed some issues/points and that's why I didn't focus on this. Tested again now here are my questions:

- that doesn't include any information about the client session. I mean `session_vars.get_map(rei)` the rei object that is used to fetch some information like user_name, client id etc.
- it is continuously being triggered (I guess because of the delay server), so hard to use for a policy
- I could not find a way to access to the collection name (again what Terrell mentioned above)
- I found the `collection` key in rei but it is empty as shown below.


```
def pep_database_gen_query_pre(rule_args, callback, rei):
    rule_exec_info = session_vars.get_map(rei)#["connection"]["option"]
    iprint(f"rei is: {rule_exec_info}", callback)
```

I guess the result of the delay server invokes:
`{"processname": "irods", "groupname": "irods", "pid": "127", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = rei is: {'plugin_instance_name': '', 'status': 0, 'operation_type': None, 'connection': None, 'data_object': None, 'collection': None, 'client_user': None, 'proxy_user': None, 'other_user': None, 'key_value_pairs': None}\n", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 38680, "server_timestamp": "2024-07-16T14:05:11.838Z", "server_type": "agent", "server_zone": "u0137480"}}`

the result when invoked by a rods user
`{"processname": "irods", "groupname": "irods", "pid": "127", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "info", "log_message": "writeLine: inString = rei is: {'plugin_instance_name': '', 'status': 0, 'operation_type': None, 'connection': None, 'data_object': None, 'collection': None, 'client_user': None, 'proxy_user': None, 'other_user': None, 'key_value_pairs': None}\n", "request_api_name": "GEN_QUERY_AN", "request_api_number": 702, "request_api_version": "d", "request_client_user": "mustafa", "request_host": "10.0.2.100", "request_proxy_user": "mustafa", "request_release_version": "rods4.3.2", "server_host": "dev-u0137480.irods.t.icts.kuleuven.be", "server_pid": 38804, "server_timestamp": "2024-07-16T14:10:14.630Z", "server_type": "agent", "server_zone": "u0137480"}}`

**Comment by korydraughn** [comment]
Hmm, there's GenQuery1, GenQuery2, and SpecificQueries. Any of them could be used by an API or client (directly) to list the contents of a collection. That makes it difficult to give you a definitive answer. But, let's say you've discovered how to detect when a collection's contents is being listed. What kinds of things are you looking to do using that information?

**Comment by mstfdkmn** [comment]
I think non of the matching PEPs for those doesn't meet the need (either not being triggered, or not able to access to info needed or it doesn't contain that info).

> What kinds of things are you looking to do using that information?

I am trying to achieve a use case that includes some details in a much simple and robust way. The use case and where I want to use that information is:

- Assume all activities to a collection need to be blocked for a while,
- For this, metadata tag will be used - when it is enabled the said collection will be locked (will not accept any activity on it and inside it)
- I was thinking to use a pep_<query>_post that will lock (return an irods error with an msi) the collection so that users during the specified time frame (the situation of metadata tag "true") would only be able to list/query that collection but they would not be able to change status (nor change to collection or its subitems like write/delete/move etc.)
- There seems to be available at least 2 more ways but they don't meet everything completely. And if [this](https://github.com/irods/irods/issues/6769) would have worked, then I guess I would use the ACLs.

**Comment by trel** [comment]
So you want to:
- prevent reading and writing of data objects 'below' a particular collection... and
- prevent creation and deletion and renames of any subcollections of that collection... 
- but allow queries into that collection

Is that correct?

**Comment by mstfdkmn** [comment]
Yes mostly correct, but in fact I want all activities (acl changes/metadata operations/read/write/rename/delete/copy for not only sub collections but all objects inside that collection) be blocked.

**Comment by trel** [comment]
You want to control metadata changes with a metadata element?  This becomes 'admin only' very quickly.

**Comment by mstfdkmn** [comment]
Yes partially, in the policy implementation there will be privileged users who will be excluded.

**Comment by mstfdkmn** [comment]
Okay, for now, I will probably focus on other two ways (block each action specifically or workaround with acls) I mentioned above for this policy implementation.

However capturing query info via PEPs might be something we would need in the last step of our auditing tool. That is, if we would like to keep track of query activities, we may need to know a correct pep that will be triggered by all sort of query actions (like ils, walk, genquery(s) etc.)
